streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamKStreamRightJoin.java

```diff
@@ -0,0 +1,221 @@
+
+        private void emitNonJoinedOuterRecords(
+                final KeyValueStore<TimestampedKeyAndJoinSide<K>, LeftOrRightValue<V2, V1>> store,
+                final Record<K, V1> record) {
+
+            // calling `store.all()` creates an iterator what is an expensive operation on RocksDB;
+            // to reduce runtime cost, we try to avoid paying those cost
+
+            // only try to emit left/outer join results if there _might_ be any result records
+            if (sharedTimeTracker.minTime + joinAfterMs + joinGraceMs >= sharedTimeTracker.streamTime) {
+                return;
+            }
+            // throttle the emit frequency to a (configurable) interval;
+            // we use processing time to decouple from data properties,
+            // as throttling is a non-functional performance optimization
+            if (internalProcessorContext.currentSystemTimeMs() < sharedTimeTracker.nextTimeToEmit) {
+                return;
+            }
+
+            // Ensure `nextTimeToEmit` is synced with `currentSystemTimeMs`, if we dont set it everytime,
+            // they can get out of sync during a clock drift
+            sharedTimeTracker.nextTimeToEmit = internalProcessorContext.currentSystemTimeMs();
+            sharedTimeTracker.advanceNextTimeToEmit();
+
+            // reset to MAX_VALUE in case the store is empty
+            sharedTimeTracker.minTime = Long.MAX_VALUE;
+
+            try (final KeyValueIterator<TimestampedKeyAndJoinSide<K>, LeftOrRightValue<V2, V1>> it = store.all()) {
+                TimestampedKeyAndJoinSide<K> prevKey = null;
+
+                boolean outerJoinLeftWindowOpen = false;
+                boolean outerJoinRightWindowOpen = false;
+                while (it.hasNext()) {
+                    if (outerJoinLeftWindowOpen && outerJoinRightWindowOpen) {
+                        break; // there are no more candidates to emit on left-outerJoin-side and right-outerJoin-side
+                    }
+                    final KeyValue<TimestampedKeyAndJoinSide<K>, LeftOrRightValue<V2, V1>> next = it.next();
+                    final TimestampedKeyAndJoinSide<K> timestampedKeyAndJoinSide = next.key;
+                    final LeftOrRightValue<V2, V1> value = next.value;
+                    final K key = timestampedKeyAndJoinSide.getKey();
+                    final long timestamp = timestampedKeyAndJoinSide.getTimestamp();
+                    sharedTimeTracker.minTime = timestamp;
+
+                    // Skip next records if window has not closed
+                    // TODO: The timestampedKeyAndJoinSide contains values from left side.
+                    final long outerJoinLookBackTimeMs = getOuterJoinLookBackTimeMs(timestampedKeyAndJoinSide);
+                    if (sharedTimeTracker.minTime + outerJoinLookBackTimeMs + joinGraceMs
+                            >= sharedTimeTracker.streamTime) {
+                        if (timestampedKeyAndJoinSide.getJoinSide() == JoinSide.LEFT) {
```

Comment:
A side note: This decision should ideally be removed. The reason is still there is because I am initialising the left- and right outer stores with the same name. See the 'init()' method in the parent class.

In the KStreamKStreamLeftJoin this call was not needed.